tests/test_io.py

```diff
@@ -218,6 +219,61 @@ def test_query_bad_params(mock_urlopen: unittest.mock.MagicMock):
     assert mock_urlopen.call_count == 0
 
 
+@pytest.mark.parametrize(
+    ("query_flag", "query_input", "expected_output"),
+    [
+        # All parameters should  not send query
+        (
+            "instrument",
+            "badInput",
+            "Not a valid instrument, please choose from "
+            + ", ".join(imap_data_access.VALID_INSTRUMENTS),
```

Comment:
No need to update here, but as an FYI in case it simplifies things later for you. `pytest.raises(match=x)` will do partial matches for you. So often I'll just put the first few words down that are what matters rather than the exact string which can get cumbersome.

```suggestion
            "Not a valid instrument"
```
and your ones later would be like "not a valid start date" because the start / end / instrument are meaningful to your message, but the exact instruments not so much.